ssm@Severga/files/ssm@Severga/applet.js

```diff
@@ -0,0 +1,452 @@
+		if (this.timeout2) {
+			Mainloop.source_remove(this.timeout2);
+			this.timeout2 = 0;
+		}
+	},
+	
+	_remove_timeoutR: function() {
+		if (this.timeoutR) {
+			Mainloop.source_remove(this.timeoutR);
+			this.timeoutR = 0;
+		}
+	},
+	
+	_fetch_cpu_temp: function(output) {
+		this._remove_timeoutR();
+		if (typeof output !== "undefined" && output != "") {
+			let lines = output.split("\n");
+			let maxTemp = 0;
+			for (let temp, i = 0; i < lines.length; i++) {
+				if (lines[i].indexOf("Package") > -1) {
+					temp = parseInt(lines[i].split(":", 2)[1].split(".", 1)[0]);
+					if (temp > maxTemp) maxTemp = temp;
+				}
+			}
+			this.currentTemp = maxTemp;
+		} else {
+			this.currentTemp = -283;
+		}
+		this.waitingForSensors = false;
+	},
+	
+	_update: function() {
+		let shortInfo, shortInfo2, longInfo, tooltip;
+		[shortInfo, longInfo] = this.cpuDataProvider.data();
+		this.cpuLabel.set_text(shortInfo);
+		tooltip = "<b>CPUs</b>   " + (this.sensorsFailed || this.currentTemp == -283 ? "" : "(<i>" + this.currentTemp + "°C max</i>)") + "\n" + longInfo + "\n\n";
+		[shortInfo, longInfo] = this.memDataProvider.data();
+		this.memLabel.set_text(shortInfo);
+		tooltip += "<b>Memory</b> (<i>" + (this.memDataProvider.memSize / 1000000000).toFixed(1) + "GB" + (this.memDataProvider.swapSize ? "; " + (this.memDataProvider.swapSize / 1000000000).toFixed(1) + "GB swap" : "") + "</i>)\n" + longInfo +"\n\n";
+		[shortInfo, shortInfo2, longInfo] = this.netDataProvider.data(this._netBlackList);
+		this.netLabel.set_text("↑" + shortInfo + "\n↓" + shortInfo2);
+		tooltip += "<b>Network</b>" + longInfo;
+		this._applet_tooltip.set_markup(tooltip);
+		let coldColor = Clutter.Color.from_string(this.coldColor); if (coldColor[0]) coldColor = coldColor[1];
+		let hotColor = Clutter.Color.from_string(this.hotColor); if (hotColor[0]) hotColor = hotColor[1];
+		let color = coldColor.interpolate(hotColor, (this.currentTemp <= this.coldTemp ? 0 : (this.currentTemp >= this.hotTemp ? 1 : (this.currentTemp - this.coldTemp) / (this.hotTemp - this.coldTemp))));
+		let r = color.red.toString(16); if (r.length == 1) r = "0" + r;
+		let g = color.green.toString(16); if (g.length == 1) g = "0" + g;
+		let b = color.blue.toString(16); if (b.length == 1) b = "0" + b;
+		this.cpuLabel.set_style("text-align: right; color: #" + r + g + b);
+		this._remove_timeout();
+		this.timeout = Mainloop.timeout_add_seconds(this.refresh, () => this._update());
+	},
+	
+	_update2: function() {
+		if (this.sensorsPath && !this.waitingForSensors && !this.sensorsFailed) {
+			this.waitingForSensors = true;
+			Util.spawn_async([this.sensorsPath], Lang.bind(this, this._fetch_cpu_temp));
+			this.timeoutR = Mainloop.timeout_add_seconds(60, () => {
+				this._remove_timeoutR();
+				this.sensorsFailed = true;
+				this.waitingForSensors = false;
+			});
+		}
+		this._remove_timeout2();
+		this.timeout2 = Mainloop.timeout_add_seconds(this.tempRefresh, () => this._update2());
```

Comment:
A suggestion on when you have recurring timers:

Instead of starting a new timer at the end of the callback each time, you can simply return `true` (or, for more readability, `GLib.SOURCE_CONTINUE`). When you no longer want the timer to continue, returning `false` (or `GLib.SOURCE_REMOVE`) will remove it (or what you do in `remove_timeout2()` is fine also).

This is fine how you have it, I just wanted to point it out.

Reply:
Thanks for the advice! I really appreciate it!